src/types.jl

```diff
@@ -226,36 +237,42 @@ const StatusFinished = Status{:finished}()
 # - onestep!
 # - trialstep!, errorcontrol! and accept!
 
+const _notdone = false
+const _done = true
+
 """
 
-Determines wheter we can take one step.  This is the core function to
-be implemented by a solver.  Note that adaptive solvers may want to
-implement only some of the substeps.
+Take a step, modifies `state` in-place.  This is the core function to
+be implemented by a solver.  However, if possible solvers should opt
+to implement the sub-step functions `trialstep!`, `errorcontrol!` and
+`accept!`, instead of directly `onestep!`.
 
-"""
+Input:
+
+- sol::Solver, state::AbstractState
 
-# onestep! tries to find out if the next step can be made
+Output:
+
+- Bool: `false`: continue iteration, `true`: terminate iteration.
+
+substeps.
+"""
 function onestep!(sol::Solver, state::AbstractState)
     opt = sol.stepper.options
     while true
         status = trialstep!(sol, state)
-
-        if !successful(status)
-            return status
-        else
-            err, statuserr = errorcontrol!(sol, state)
-            if !successful(statuserr)
-                return statuserr
-            elseif err <= 1
-                statusaccept = accept!(sol, state)
-                if !successful(statusaccept)
-                    return statusaccept
-                else
-                    return status
-                end
-            end
+        err, status = errorcontrol!(sol, state, status)
```

Comment:
And this would bring more needless complexity to the `errorcontrol!`.  Unless you want to use `status` to generate new stepsize or whatever.  But this is completely internal to the solver and could use `state`.